File: src/main/java/io/binac/leetcode/MaximalRectangle.java
package io.binac.leetcode;

import java.util.Arrays;

/**
 * Given a 2D binary matrix filled with 0's and 1's, find the largest rectangle containing only 1's and return its area.
 * <p>
 * <p>Example:
 * <blockquote><pre>
 *     Input:
 *     [
 *       ["1","0","1","0","0"],
 *       ["1","0","1","1","1"],
 *       ["1","1","1","1","1"],
 *       ["1","0","0","1","0"]
 *     ]
 *     Output: 6
 * </blockquote></pre>
 */
public class MaximalRectangle {
    public static class Solution1 {
        public int maximalRectangle(char[][] matrix) {
            final int row = matrix.length;
            if (row == 0) return 0;
            final int col = matrix[0].length;

            int start[] = new int[col], end[] = new int[col], height[] = new int[col];
            Arrays.fill(end, col);
            int max = 0;
            for (int i = 0; i < row; ++i) {
                int s = 0;
                for (int j = 0; j < col; ++j) {
                    if (matrix[i][j] == '1') {
                        start[j] = Math.max(start[j], s);
                        ++height[j];
                    } else {
                        start[j] = 0;
                        s = j + 1;
                        height[j] = 0;
                    }
                }
                int e = col;
                for (int j = col - 1; j >= 0; --j) {
                    if (matrix[i][j] == '1') {
                        end[j] = Math.min(end[j], e);
                        max = Math.max(max, height[j] * (end[j] - start[j]));
                    } else {
                        end[j] = col;
                        e = j;
                    }
                }
            }
            return max;
        }
    }
}

File: python3/leetcodepy/maximal_rectangle.py
"""
Given a rows x cols binary matrix filled with 0's and 1's, find the largest rectangle containing only 1's and return its area.



Example 1:
https://assets.leetcode.com/uploads/2020/09/14/maximal.jpg

Input: matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
Output: 6
Explanation: The maximal rectangle is shown in the above picture.

Example 2:

Input: matrix = [["0"]]
Output: 0

Example 3:

Input: matrix = [["1"]]
Output: 1


Constraints:

rows == matrix.length
cols == matrix[i].length
1 <= row, cols <= 200
matrix[i][j] is '0' or '1'.
"""
from typing import List


class Solution1:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        heights = [0] * n
        left = [0] * n
        right = [n] * n
        result = 0
        for i in range(m):
            cur_left = 0
            for j in range(n):
                if matrix[i][j] == '0':
                    left[j] = 0
                    cur_left = j + 1
                    heights[j] = 0
                else:
                    left[j] = max(left[j], cur_left)
                    heights[j] += 1
            cur_right = n
            for j in range(n - 1, -1, -1):
                if matrix[i][j] == '0':
                    right[j] = n
                    cur_right = j
                else:
                    right[j] = min(right[j], cur_right)
                    result = max(result, heights[j] * (right[j] - left[j]))
        return result


class Solution2:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        heights = [0] * n
        result = 0
        for i in range(m):
            stack = []
            for j in range(n):
                heights[j] = 0 if matrix[i][j] == '0' else heights[j] + 1
                while stack and heights[j] <= heights[stack[-1]]:
                    k = stack.pop()
                    left = 0 if not stack else stack[-1] + 1
                    result = max(result, (j - left) * heights[k])
                stack.append(j)
            while stack:
                k = stack.pop()
                left = 0 if not stack else stack[-1] + 1
                result = max(result, (n - left) * heights[k])
        return result
